src/katgpucbf/fgpu/engine.py

```diff
@@ -419,19 +419,68 @@ def __init__(self, output: Output, engine: "Engine", context: AbstractContext, s
         self.gains = np.zeros((output.channels, self.pols), np.complex64)
+        self._populate_sensors()
         self._init_delay_gain()
 
         self.descriptor_heap = send.make_descriptor_heap(
             channels_per_substream=output.channels // len(substreams),
             spectra_per_heap=engine.spectra_per_heap,
         )
 
+    def _populate_sensors(self) -> None:
+        sensors = self.engine.sensors
+        for pol in range(N_POLS):
+            sensors.add(
+                aiokatcp.Sensor(
+                    str,
+                    f"{self.output.name}.input{pol}.eq",
+                    "For this input, the complex, unitless, per-channel digital scaling factors "
+                    "implemented prior to requantisation",
+                    initial_status=aiokatcp.Sensor.Status.NOMINAL,
+                )
+            )
+            sensors.add(
+                aiokatcp.Sensor(
+                    str,
+                    f"{self.output.name}.input{pol}.delay",
+                    "The delay settings for this input: (loadmcnt <ADC sample "
+                    "count when model was loaded>, delay <in seconds>, "
+                    "delay-rate <unit-less or, seconds-per-second>, "
+                    "phase <radians>, phase-rate <radians per second>).",
+                )
+            )
+            # TODO[nb]: it may be better to make this a single sensor and
+            # have only one of the pipelines update it.
+            sensors.add(
+                aiokatcp.Sensor(
+                    float,
+                    f"{self.output.name}.input{pol}.dig-pwr-dbfs",
+                    "Digitiser ADC average power",
+                    units="dBFS",
+                    status_func=dig_pwr_dbfs_status,
+                    auto_strategy=aiokatcp.SensorSampler.Strategy.EVENT_RATE,
+                    auto_strategy_parameters=(MIN_SENSOR_UPDATE_PERIOD, math.inf),
+                )
+            )
+            sensors.add(
+                aiokatcp.Sensor(
+                    int,
+                    f"{self.output.name}.input{pol}.feng-clip-cnt",
+                    "Number of output samples that are saturated",
+                    default=0,
+                    initial_status=aiokatcp.Sensor.Status.NOMINAL,
+                    auto_strategy=aiokatcp.SensorSampler.Strategy.EVENT_RATE,
+                    auto_strategy_parameters=(MIN_SENSOR_UPDATE_PERIOD, math.inf),
+                )
+            )
+
     def _init_delay_gain(self) -> None:
         """Initialise the delays and gains."""
-        # TODO[nb]: create per-pipeline sensors
         for pol in range(N_POLS):
             delay_model = MultiDelayModel(
-                callback_func=partial(self.update_delay_sensor, delay_sensor=self.engine.sensors[f"input{pol}.delay"])
+                callback_func=partial(
+                    self.update_delay_sensor, delay_sensor=self.engine.sensors[f"{self.output.name}.input{pol}.delay"]
```

Comment:
I don't really like this in-between thing that `black` does, to me it makes more sense to either have the entire function call on one line, or each argument on its own line.

No change needed, just felt the need to complain about this a bit.

Reply:
All the line-wrapping in this function is a symptom that it's trying to stuff too much into a single statement. See if you like the new version better.